etl/lib/dbsessions/mongodb.py

```diff
@@ -2,52 +2,69 @@
 import json
 from pandas import DataFrame
 import re
-from tg.i18n import ugettext as _
 from tg.exceptions import HTTPPreconditionFailed
 
 
 class MongoDBSession(object):
+    _session = None
+    _collection = None
 
     def __init__(self, url):
+        collection = None
+        if u'#' in url:
+            url, collection = url.rsplit('#', 1)
+        self.set_session(url)
+        if collection is not None:
+            self.set_collection(collection)
+
+    def set_session(self, url):
         from pymongo import MongoClient
         self._session = MongoClient(url)
-        self._collection = None
 
     def set_collection(self, collection):
-        if collection in self._session.get_default_database().list_collection_names():
-            self._collection = self._session.get_default_database() \
-                .get_collection(collection)
-        else:
-            raise HTTPPreconditionFailed(
-                detail='The requested collection: %s doesn\'t exist, check it' % collection
-            )
+        self._collection = self._session.get_default_database() \
+            .get_collection(collection)
 
     def extract_directive_value(self, query, directive_pattern):
-        pattern = r'^#' + re.escape(directive_pattern) + r'.*'
+        pattern = r'^[\s]*#' + re.escape(directive_pattern) + r'.*' + re.escape('=') + r'.*'
         directive_match = re.search(pattern, query)
         if directive_match:
             value = directive_match.group(0).split('=')[-1].strip()
             return value, directive_match
+        elif not directive_match and self._collection is not None:
+            return None, None
         else:
             raise HTTPPreconditionFailed(
-                detail=_('Wrong collection directive syntax, it should be something like "#collection=collectionname"')
+                detail=
+                'Missing collection, it should be something like "#collection=collectionname" in the query or'
+                'specify the url in form mongodb://user:pass@host:port/database#collection'
             )
 
     def parse_q(self, q):
         collection_name, sub_re = self.extract_directive_value(q, 'collection')
```

Comment:
`parse_q` seems a bit overcomplex and supports a single directive at a time.

Seems parsing the lines would be far simpler to understand and I'm sure there are way to simplify further:
```
def parse_query(query):
    q_parts = []
    directives = {}
    for line in query.split('\n'):
        stripped_line = line.strip()
        if stripped_line[:1] == "#":
            key, value = [v.strip() for v in stripped_line.split("=", 1)]
            directives[key.strip('#').strip()] = value
        else:
            q_parts.append(line)
    q = '\n'.join(q_parts).strip()
    return directives, q
```
also supports any number of directives

```
>>> parse_query("""
... # collection = hello
...  # foobar = something else
... {"name": "Hello",
...  "surname": "World"}
""")
({' foobar': 'something else', ' collection': 'hello'}, '{"name": "Hello",\n "surname": "World"}')
```